// @webex/plugin-meetings/dist/ciscospark.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

require('@ciscospark/plugin-authorization');

require('@ciscospark/internal-plugin-wdm');

require('@ciscospark/plugin-logger');

require('@ciscospark/plugin-people');

require('@ciscospark/plugin-rooms');

var _sparkCore = require('@ciscospark/spark-core');

var _sparkCore2 = _interopRequireDefault(_sparkCore);

var _merge = require('lodash/merge');

var _merge2 = _interopRequireDefault(_merge);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CiscoSpark = _sparkCore2.default.extend({
  ciscospark: true,
  version: '1.50.6'
});

CiscoSpark.version = '1.50.6';

CiscoSpark.init = function init() {
  var attrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var config = (0, _merge2.default)({}, {
    hydra: process.env.HYDRA_SERVICE_URL || 'https://api.ciscospark.com/v1',
    credentials: {
      clientType: 'confidential'
    },
    device: {
      ephemeral: true
    },
    storage: {
      boundedAdapter: _sparkCore.MemoryStoreAdapter,
      unboundedAdapter: _sparkCore.MemoryStoreAdapter
    }
  }, attrs.config);
  return new CiscoSpark((0, _merge2.default)({}, { config: config }, attrs));
};

exports.default = CiscoSpark;
