// @webex/plugin-meetings/dist/meeting-info/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _sparkCore = require('@ciscospark/spark-core');

var _constants = require('../constants');

var _collection = require('./collection');

var _collection2 = _interopRequireDefault(_collection);

var _request = require('./request');

var _request2 = _interopRequireDefault(_request);

var _util = require('./util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class MeetingInfo
 */
var MeetingInfo = function (_StatelessSparkPlugin) {
  (0, _inherits3.default)(MeetingInfo, _StatelessSparkPlugin);

  /**
   *
   * @param {Object} attrs
   * @param {Object} options
   */
  function MeetingInfo(attrs, options) {
    (0, _classCallCheck3.default)(this, MeetingInfo);

    var _this = (0, _possibleConstructorReturn3.default)(this, (MeetingInfo.__proto__ || (0, _getPrototypeOf2.default)(MeetingInfo)).call(this, {}, options));

    _this.namespace = _constants.MEETINGS;

    _this.meetingInfoRequest = new _request2.default({}, options);
    return _this;
  }

  /**
   * @param {string} id
   * @returns {object}
   * @memberof MeetingInfo
   */


  (0, _createClass3.default)(MeetingInfo, [{
    key: 'requestFetchInfo',


    /**
     * Helper to make the actual MeetingInfo request and set the meetingInfo if successful, else reject
     * @param {Object} options
     * @returns {Promise}
     * @private
     */
    value: function requestFetchInfo(options) {
      return this.meetingInfoRequest.fetchMeetingInfo(options).then(function (info) {
        MeetingInfo.setMeetingInfo(info.id, info);
        return info;
      }).catch(function (error) {
        return _promise2.default.reject(error);
      });
    }

    /**
     * Helper to generate the options for the MeetingInfo request
     * @param {String} destination
     * @param {String} type
     * @returns {Object}
     * @private
     */

  }, {
    key: 'fetchInfoOptions',
    value: function fetchInfoOptions(destination, type) {
      return _util2.default.generateOptions({
        destination: destination,
        type: type,
        conversationServiceUrl: this.spark.internal.device.services.conversationServiceUrl
      });
    }

    /**
     * Fetches meeting info from the server
     * @param {String} destination one of many different types of destinations to look up info for
     * @param {String} type optional type to match up with the destination value
     * @returns {Promise} returns a meeting info object
     * @public
     */

  }, {
    key: 'fetchMeetingInfo',
    value: function fetchMeetingInfo(destination) {
      var _this2 = this;

      var type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

      var options = this.fetchInfoOptions(destination, type);
      // fetch meeting info
      return this.requestFetchInfo(options).catch(function (error) {
        // if it failed the first time as meeting link
        if (options.type === _constants.MEETING_LINK) {
          // convert the meeting link to sip URI and retry
          return _this2.requestFetchInfo(_this2.fetchInfoOptions(_util2.default.convertLinkToSip(destination), _constants.SIP_URI));
        }
        return _promise2.default.reject(error);
      });
    }
  }], [{
    key: 'getMeetingInfo',
    value: function getMeetingInfo(id) {
      return _collection2.default.get(id);
    }

    /**
     * @param {string} id
     * @param {object} info
     * @returns {null}
     * @memberof MeetingInfo
     */

  }, {
    key: 'setMeetingInfo',
    value: function setMeetingInfo(id, info) {
      _collection2.default.set(id, info);
    }
  }]);
  return MeetingInfo;
}(_sparkCore.StatelessSparkPlugin); /*!
                                     * Copyright (c) 2015-2018 Cisco Systems, Inc. See LICENSE file.
                                     */

exports.default = MeetingInfo;
